### ISSUE TITLE
[8.0-preview2] error MSB4012: The expression "CompileListCount=@(_ReadyToRunCompileList->Count())" cannot be used in this context

### ISSUE THREAD

**Issue by MichaelSimons** [dotnet/runtime#82795]
The bootstrap leg of source-build is failing in 8.0 preview2 with the following error:

/repos/dotnet2/src/runtime/artifacts/source-build/self/src/artifacts/bin/Crossgen2Tasks/Debug/net8.0/Microsoft.NET.CrossGen.targets(349,49): error MSB4012: The expression "CompileListCount=@(_ReadyToRunCompileList->Count())" cannot be used in this context. Item lists cannot be concatenated with other strings where an item list is expected. Use a semicolon to separate multiple item lists. [/repos/dotnet2/src/runtime/artifacts/source-build/self/src/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.Composite.sfxproj]

[Binlog](https://github.com/MichaelSimons/Sandbox/raw/master/stage2.runtime.sourcebuild.zip) 

This is a blocker for the preview2 release.  Any help in diagnosing would be appreciated.  TIA

cc @dotnet/crossgen-contrib

**Comment by trylek** [comment]
Hmm, it looks like @davidwrighton was the last person touching this project in

https://github.com/dotnet/runtime/pull/52579

but it's not clear to me from the PR whether the problematic line

https://github.com/dotnet/runtime/blob/5e3dd8da34f1ffc3032197b3f9d52797431391c5/src/tasks/Crossgen2Tasks/Microsoft.NET.CrossGen.targets#L349

was authored by him or not. If I understand the error message properly, MSBuild has a valid point that such a compound argument cannot be unambiguously parsed as the semicolon separator is used in two different contexts - to separate key=value pairs and multiple values. I have no idea how that was supposed to work.

**Comment by MichaelSimons** [comment]
For clarity in the bootstrap build, we are using a source-built preview 2 sdk/toolset to rebuild source-build with.  It is possible this is a new msbuild check.

**Comment by rainersigwald** [comment]
This is not a new check and I don't think we've touched it recently.

**Comment by rainersigwald** [comment]
It's firing now because of https://github.com/dotnet/sdk/pull/30269. @baronfel, that puts MSBuild in "you must want a list of items" mode, so an item transform complains when there's stuff around it. In this specific case `preamble_@(Foo->Count())_suffix` the concatenation is completely fine and wouldn't result in problems but our analysis isn't deep enough to see that.

**Comment by MichaelSimons** [comment]
@rainersigwald - is there a workaround we can apply in source-build to unblock us while a long term solution is discussed/implemented?

**Comment by trylek** [comment]
I'm glad to hear that, while I apparently completely misdiagnosed the issue, I have managed to pull in the right people :-).

**Comment by trylek** [comment]
@rainersigwald / @baronfel - the PR https://github.com/dotnet/sdk/pull/30269 is apparently against a .NET 7 servicing branch unless I'm mistaken, do you know whether the equivalent fix has been merged into SDK main?

**Comment by MichaelSimons** [comment]
> @rainersigwald / @baronfel - the PR https://github.com/dotnet/sdk/pull/30269 is apparently against a .NET 7 servicing branch unless I'm mistaken, do you know whether the equivalent fix has been merged into SDK main?

Yes, changes made in servicing branches automatically flow into main.  It flowed into main with https://github.com/dotnet/sdk/pull/30645.

**Comment by trylek** [comment]
For a possible workaround, if I understand Rainer's explanation correctly, it should be possible to split the construction of the <code>AllowEmptyTelemetry</code> property into several parts, in particular we should be able to isolate the <code>Count()</code> computation by using it to populate an auxiliary property we would subsequently merge with the rest.

**Comment by baronfel** [comment]
So I believe the fix here is for runtime to use this factoring of the new item-based pattern:


```xml
	<ItemGroup>
       <_R2RCrossgenTelemetry Include="PublishReadyToRunUseCrossgen2" Value="$(PublishReadyToRunUseCrossgen2)" />
       <_R2RCrossgenTelemetry Include="Crossgen2PackVersion" Value="%(ResolvedCrossgen2Pack.NuGetPackageVersion)" />
       <_R2RCrossgenTelemetry Include="CompileListCount" Value="@(_ReadyToRunCompileList->Count())" />
       <_R2RCrossgenTelemetry Include="FailedCount" Value="@(_ReadyToRunCompilationFailures->Count())" />
    </ItemGroup>
    <AllowEmptyTelemetry EventName="ReadyToRun" EventData="@(_R2RCrossgenTelemetry)" />
```

**Comment by trylek** [comment]
@baronfel - just to satisfy my curiosity, do you happen to know who is consuming the <code>AllowEmptyTelemetry</code> property? I mean, I can imagine a parser that can do <code>A=B;C;D;E=F;G</code> but it must be a pretty sophisticated parser so I'm wondering whether we might be just dropping some of these key-value items on the floor unless it's specifically aware of this peculiar detail.

**Comment by MichaelSimons** [comment]
@baronfel - thanks I am testing this now in source-build.  Can you open an SDK PR to fix this in the preview2 branch?  Also should this issue be transferred to the sdk repo?

**Comment by baronfel** [comment]
@MichaelSimons I'm not sure it's an SDK PR - the targets in question moved to the runtime repo, and all of the SDK uses of AllowEmptyTelemetry have been fixed up? 

@trylek The `AllowEmptyTelemetry` Task is a bridge to the MSBuild engine's built-in telemetry that the CLI uses to send telemetry to our AppInsights data pools. There are a bunch of other uses of it in the SDK codebase. The `A=B;C;D;E=F;G` is normal MSBuild syntax for a piece of Metadata that is secretly a list of things, and MSBuild has rules for converting to and fro. The implementation is surprisingly straightforward from the task's perspective: [AllowEmptyTelemetry](https://github.com/dotnet/sdk/blob/9f43a4a55078a22daac527c99e98d26e827a0ee2/src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs#L26).

**Comment by MichaelSimons** [comment]
> @MichaelSimons I'm not sure it's an SDK PR - the targets in question moved to the runtime repo, and all of the SDK uses of AllowEmptyTelemetry have been fixed up?

I see that now.  Thanks for clarifying for me.

**Comment by baronfel** [comment]
I opened https://github.com/dotnet/runtime/pull/82804 against Runtime with the same change, is main enough or is there a better branch to target?

**Comment by MichaelSimons** [comment]
Main and ported to preview 2.

**Comment by MichaelSimons** [comment]
FYI I verified the fix works in source-build.  Thanks for the quick help to unblock preview 2.

**Comment by MichaelSimons** [comment]
I think this can be closed now as a fix was checked into main.
